content/post/the-lab-book.md

```diff
@@ -0,0 +1,87 @@
+
+When it comes to understanding *_why_* a given model is structured the way it is, you need a history of what was already tried and what worked and what didn’t. If someone on the team wants to pick up where the first practitioner left off, they can avoid lots of wasted effort if they know what was attempted before.
+
+## When to use the lab book?
+
+Have you ever found yourself in a situation where you are working on a gnarly piece of code and you feel like you are just slamming your head into a brick wall over and over and over again? This happens to me pretty frequently and, if you’re around me while I’m going through this, you’ll hear me repeatedly say, aloud, “how on earth could *_that_* be happening?”. At about the third time I say that phrase aloud, I know I need to bust out the lab book.
+
+Software complexity (especially in the data science space) can make it difficult or impossible to reason about what’s happening. For example:
+
+* Working third-party opaque libraries
+* Stochastic processes
+* “Black-box” ML models
+* Any piece of software sufficiently complex that you can’t hold all of the state in your head at a given time
+
+At any point where you are working on a task and you are repeatedly encountering unexpected or surprising results that’s a good time to use the data science lab book to begin keeping track of your hypotheses, the evidence you have learned, and the assumptions you’re making.
+
+In particular, it’s a good idea to use a lab book when:
+
+* You’re doing experimental research (e.g., model-building and model-fitting)
+* You’ve encountered anomalous patterns in results that haven’t gone away with 3 different attempts
+* You’re isolating a bug in an opaque part of the code-base
+
+By recording your experiences, you’ll not only force yourself to think more clearly and carefully about the problem at hand, but you’ll also be building a record that your colleagues can review when they either join you to help out or want to learn from your process.
+
+## What does the lab book for data science look like?
+
+A lab book should have one entry per experiment and should be ordered in reverse-chronological order (newest experiments at the top). A google-doc works great for this, especially if you’re collaborating with other team-members.
+
+Here’s an example template:
+
+**Date:** 2019-09-15  
+**Hypothesis:** Our seasonality specification is sub-optimal because we don’t have sufficient data to estimate it. We can use google-trends data as an external data source for seasonality.  
+**Conditions:** Included google trends data (scaled and centered) for keywords “cold” and “flu” in the regression component of the model.  
+**Results:** Predictive accuracy improved by 2.5% but the model took ~30 minutes longer to converge. Unclear why.  
+**Git commit:** b676ff0747e56fe3c2df30c3e93bb5f097289291
```

Comment:
I think it would be really interesting to talk a bit more about how the lab notebook can work best with Git. Are there any practices you'd recommend implementing in Git to make it more compatible with lab notebook record keeping?

Reply:
Good call -- adding!